csv_permissions/permissions.py

```diff
@@ -182,168 +64,253 @@ def _parse_csv(
                 continue
-            model_name_orig = row["Model"]  # note that capitalisation may differ to model._meta.model_name
 
-            if any(model_name_orig.strip().startswith(comment_prefix) for comment_prefix in ("//", "#")):
+            if any(row[0].strip().startswith(comment_prefix) for comment_prefix in ("//", "#")):
                 # Ignore lines beginning with comment chars
                 continue
 
-            app_config = apps.get_app_config(row["App"])
+            if len(row) < prelim_header_count:
+                raise ValueError(f"Incomplete line {line_number} in {csv_file}")
+
+            # note that model capitalisation may differ to model._meta.model_name
+            model_name_orig, app_label, action, is_global = row[:prelim_header_count]
+
+            app_config = apps.get_app_config(app_label)
             model = app_config.get_model(model_name_orig) if model_name_orig else None
 
-            action = row["Action"]
-            if row["Is Global"] == "yes":
+            if is_global == "yes":
                 is_global = True
-            elif row["Is Global"] == "no":
+            elif is_global == "no":
                 is_global = False
             else:
-                raise RuntimeError("Invalid value for Is Global.")
-
-            if _PREDEFINED_ACTION_IS_GLOBAL.get(action, is_global) != is_global:
-                raise RuntimeError(
-                    "Invalid action / global setting for"
-                    f" {app_config.label}.{model_name_orig}.{action} (is_global should"
-                    f" not be {is_global})"
-                )
-
-            rule_name = CSV_PERMISSIONS_RESOLVE_RULE_NAME(app_config, model, action, is_global)
-
-            if rule_name not in perm_is_global:
-                perm_is_global[rule_name] = is_global
-                perm_group_details[rule_name] = {}
-
-            for auth_group in auth_groups:
-                access_level = row[auth_group]
-
-                perm_group_details[rule_name][auth_group] = PartiallyResolvedPermission(
-                    app_config=app_config,
-                    model=model,
-                    action=action,
-                    access_level=access_level,
-                )
+                raise ValueError("Invalid value for Is Global.")
+
+            permission = resolve_permission_name_func(app_config, model, action, is_global)
+
+            if permission not in perm_is_global:
+                perm_is_global[permission] = is_global
+                perm_user_type_unresolved[permission] = {}
+
+            for i, user_type in enumerate(user_type_headers):
+                try:
+                    evaluator_name = row[prelim_header_count + i]
+                except IndexError:
+                    continue
+
+                if user_type == "":
+                    # if a column has an empty user type then that's allowed but only if the entire column is empty
```

Comment:
what's the reason for an empty column?

Reply:
second this question, find it more confusing than helpful (harder for a household cat)

Reply:
If you have trailing commas in your CSV file then the last column will be read as an empty header.

It was only slightly more work to then allow empty columns in the middle of the file too, but means that you can break up your users if you want (eg in the current project I'm working on we have 3 user profiles, each with 2-3 user_types in each profile so leaving a blank column can make it a little easier to visually break them up)